components/support/uniffi/README.md

```diff
@@ -0,0 +1,204 @@
+with field access and serialization helpers. Again, much like we currently do with protobufs.
+
+When passing back and forth over the FFI, records are serialized to a byte buffer and
+deserialized on the other side. We could continue to use protobuf for this, but I suspect
+we could come up with a simpler scheme given we control both sides of the pipe. Callers
+passing a record must keep the buffer alive until the callee returns; callers receiving
+a record must call a destructor to free the buffer after hydrating an object on their side.
+
+### Sequences
+
+Both WebIDL and WIT have a builtin `sequence` type and we should use it verbatim.
+
+```
+interface MyObject {
+    sequence<Foo> getAllTheFoos();
+}
+```
+
+We currently use ad-hoc Protobuf messages for this, e.g. the `AccountEvent` and
+`AccountEvents` types in fxa-client. But it seems reasonable to support a generic
+implementation on both sides of the FFI boundary.
+
+When traversing the FFI, these would be serialized into a byte buffer and parsed
+back out into a Vec or Array or whatever on the other side. Just like Record types.
+
+### Enums
+
+WebIDL as simple C-style enums, like this:
+
+```
+enum AccountEventType {
+  "INCOMING_DEVICE_COMMAND",
+  "PROFILE_UPDATED",
+  "DEVICE_CONNECTED",
+  "ACCOUNT_AUTH_STATE_CHANGED",
+  "DEVICE_DISCONNECTED",
+  "ACCOUNT_DESTROYED",
+};
+```
+
+These could be encoded into an unsigned integer type for transmission over the FFI,
+like the way we currently handle error numbers.
+
+There is also more sophisticated stuff in there, like union types and nullable
+types. I haven't really thought about how to map those on to what we need.
+
+### Callbacks
+
+WebIDL has some syntax for them, but I haven't looked at this in any detail
+at all. It seems hard, but also extremely valuable because handling callbacks
+across the FFI boundary has been a pain point for us in the past.
+
+## Scaffolding Generation
+
+Currently a very hacky attempt in [./src/scaffolding.rs](./src/scaffolding.rs),
+and a `generate_component_scaffolding(idl_file: &str)` function that's intended
+to be used from the component's build file.
+
+See the [fxa-client crate](../../fxa-client/build.rs) for an example of (attempting to)
+use this, although it's in a very incomplete state.
+
+It doesn't produce working Rust code, but it'll produce Rust-shaped kind-of-code
+that gives you a bit of an idea what it's going for.
+
+Could really benefit from a templating library rather than doing a bunch of
+`writeln!()` with raw source code fragements.
+
+## Language Bindings Generation
+
+Currently totally unimplemented.
+
+A great opportunity for anyone interested to
+dive in! You could try looking at the hand-written Kotlin or Swift code for
+the fxa-client component, and see if you can generate something similar from
+`fxa-client.idl`. Take a look at the way the scaffolding generator works to
+see how to get started.
+
+## What could possibly go wrong?
```

Comment:
This whole section is a big part of why I never pushed too hard here, especially the bit about "The resulting autogenerated code might be a lot harder to debug when things go wrong".

I am/was very worried that anything that's not very close to a perfect solution would only make things more likely to fail, and I still have that fear.

That said, I'm still very interested in this and am planning on diving in a little more when my schedule frees up a bit (there's a lot of telemetry work that needs to land with a bit of urgency, so I can't in good conscience dig in until that's in a better state)

Reply:
Aye. I think it would have been possible for us to go too far too early down this road, before we'd really lived with writing and maintaining the code by hand and understood its failure modes. I'm lightly hopeful that we've been doing this for long enough now that it's time for some automation, but I'm still pretty suspicious of hidden costs.

Reply:
Color me skeptical too - however, a possible benefit is that we will not *need* to debug the generated code. Debugging the code generated by a compiler is tricky too, but it's so rare that we never really consider whether recent compiler changes have made this less or more difficult.

IOW, we are probably suffering from stockholm syndrome :)

again, to be clear, I believe we all agree that (a) this is far from proven but (b) the benefits are such that it's worth exploring :)